server/lib/src/idm/ldap.rs

```diff
@@ -477,6 +478,94 @@ impl LdapServer {
         Ok(result)
     }
 
+    #[instrument(level = "debug", skip_all)]
+    async fn do_compare(
+        &self,
+        _idms: &IdmServer,
+        cr: &CompareRequest,
+        uat: &LdapBoundToken,
+        _source: Source,
+    ) -> Result<Vec<LdapMsg>, OperationError> {
+        admin_info!("Attempt LDAP CompareRequest for {}", uat.spn);
+
+        // convert to a search request, and then do the search.
+        let sr = SearchRequest {
+            msgid: cr.msgid,
+            base: cr.entry.clone(),
+            scope: LdapSearchScope::Base,
+            filter: LdapFilter::Equality(cr.atype.clone(), cr.val.clone()),
+            attrs: vec!["1.1".to_string()], // we only need to know if it exists.
+        };
+
+        trace!("Converting to Search: {:?}", sr);
+
+        fn compare_result_msg(cr: &CompareRequest, code: LdapResultCode) -> LdapMsg {
+            LdapMsg {
+                msgid: cr.msgid,
+                op: LdapOp::CompareResult(LdapResult {
+                    code,
+                    matcheddn: "".to_string(),
+                    message: "".to_string(),
+                    referral: vec![],
+                }),
+                ctrl: vec![],
+            }
+        }
+
+        let res = match self.do_search(_idms, &sr, uat, Source::Internal).await {
+            Ok(r) => r,
+            Err(OperationError::InvalidAttributeName(s)) => {
+                request_error!(?s, "Invalid Attribute Name in CompareRequest");
+
+                return Ok(vec![compare_result_msg(
+                    cr,
+                    LdapResultCode::UndefinedAttributeType,
+                )]);
+            }
+            Err(e) => {
+                request_error!(err = ?e, "Error doing search for compare request");
+
+                return Err(e);
+            }
+        };
+
+        if let Some((srd, results)) = &res.split_last() {
+            if let LdapOp::SearchResultDone(LdapResult {
+                code: LdapResultCode::Success,
+                ..
+            }) = &srd.op
+            {
+                if !results.is_empty() {
+                    return Ok(vec![compare_result_msg(cr, LdapResultCode::CompareTrue)]);
+                }
+
+                // search again to determine if entry exists at all
+                let mut sr = sr;
+                sr.filter = LdapFilter::Present(Attribute::Spn.to_string());
+
+                let res = self.do_search(_idms, &sr, uat, Source::Internal).await?;
```

Comment:
The problem with calling this search twice is that they are done in two separate transactions, so it's possible the entry changes from the first search to the second. As a result, we can't just make do_compare a wrapper over do_search *unless* there is a way to do it in a single search. Most likely we need to "duplicate" do_search and strip out anything that isn't needed.

Reply:
Good catch. Do you think it makes sense to duplicate this or shall I just remove the not strictly necessary second search and just return `CompareTrue` vs `CompareFalse` based on the first one?

Reply:
The problem then is you can't distinguish a compareFalse from no such entry. I think it needs to actually be a dedicated operation like do_search, we can't just layer it on top. If that's too much, I'm happy to edit the PR and fix that :)

Reply:
Alright, I will try to implement it properly. Thanks for your help 😊

Reply:
You should be able to "copy paste" the insides of do_search, and then strip out all the parts excepting base-dn searches. We could probably even refactor a few parts to be common. 

Remember also the internal query server has both search (which returns entries) and exists which gives a true or false if the condition is met. I think you want the latter :)

Reply:
I have changed the implementation accordingly. I removed quite a lot of code after the copying and I think there is not much deduplication to be achieved - but I can try. Please first let me know if you think the current implementation is correct and I will continue from there. Thanks!

Reply:
Can you also add yourself to contributors.md in the root of the project if you wish? :)

Reply:
Sure, added myself.